File: cardgame_tutorial/Ex06.java
// Ex06.java

import ch.aplu.jcardgame.*;
import ch.aplu.jgamegrid.*;

public class Ex06 extends CardGame
{
  public enum Suit
  {
    SPADES, HEARTS, DIAMONDS, CLUBS
  }

  public enum Rank
  {
    ACE, KING, QUEEN, JACK, TEN
  }
  //
  private final Deck deck =
    new Deck(Suit.values(), Rank.values(), "cover");
  private final int nbPlayers = 4;
  private final Location[] handLocations =
  {
    new Location(300, 525),
    new Location(75, 300),
    new Location(300, 75),
    new Location(525, 300),
    new Location(300, 300)
  };
  private final Location[] bidLocations =
  {
    new Location(300, 350),
    new Location(250, 300),
    new Location(300, 250),
    new Location(350, 300),
  };
  private final Location[] stockLocations =
  {
    new Location(400, 500),
    new Location(100, 400),
    new Location(200, 100),
    new Location(500, 200),
  };
  private Hand[] hands;
  private Hand[] bids = new Hand[nbPlayers];
  private Hand[] stocks = new Hand[nbPlayers];
  private Hand talon;

  public Ex06()
  {
    super(600, 600, 30);
    dealingOut();
    setStatusText("Dealing out...done. Double click to play.");
    initBids();
    initStocks();
  }

  private void dealingOut()
  {
    setStatusText("Dealing out...");

    hands = deck.dealingOut(nbPlayers, 0);  // All cards in talon
    talon = hands[nbPlayers];
    talon.setVerso(true);

    for (int i = 0; i < nbPlayers; i++)
    {
      hands[i].setView(this, new StackLayout(handLocations[i], 90 * i));
      hands[i].draw();
      hands[i].setTouchEnabled(true);
    }

    talon.setView(this, new StackLayout(handLocations[nbPlayers]));
    talon.draw();

    while (!talon.isEmpty())
    {
      for (int i = 0; i < nbPlayers; i++)
      {
        Card top = talon.getLast();
        top.setVerso(i == 0 ? false : true);
        talon.transfer(top, hands[i], true);
      }
    }
  }

  private void initBids()
  {
    for (int i = 0; i < nbPlayers; i++)
    {
      bids[i] = new Hand(deck);
      bids[i].setView(this, new StackLayout(bidLocations[i]));
      bids[i].addCardListener(new CardAdapter()
      {
        public void atTarget(Card card, Location loc)
        {
          card.getHand().draw();
        }
      });
      hands[i].setTargetArea(new TargetArea(bidLocations[i]));
      hands[i].setTouchEnabled(true);
      final int k = i;
      hands[i].addCardListener(new CardAdapter()
      {
        public void leftDoubleClicked(Card card)
        {
          card.setVerso(false);
          card.transfer(bids[k], true);
        }

        public void rightClicked(Card card)
        {
          transferToStock(k);
        }
      });
    }
  }

  private void initStocks()
  {
    for (int i = 0; i < nbPlayers; i++)
    {
      stocks[i] = new Hand(deck);
      double rotationAngle;
      if (i == 0 || i == 2)
        rotationAngle = 0;
      else
        rotationAngle = 90;
      stocks[i].
        setView(this, new StackLayout(stockLocations[i], rotationAngle));
    }
  }

  private void transferToStock(int player)
  {
    for (int i = 0; i < nbPlayers; i++)
    {
      bids[i].setTargetArea(new TargetArea(stockLocations[player]));
      Card c = bids[i].getLast();
      if (c == null)
        continue;
      c.setVerso(true);
      bids[i].transferNonBlocking(c, stocks[player], true);
    }
  }

  public static void main(String[] args)
  {
    new Ex06();
  }
}

File: python/CardGameEx/Ex06.py
# Ex06.py

from ch.aplu.jgamegrid import Location
from ch.aplu.jcardgame import Deck, Hand, CardGame, StackLayout, TargetArea, CardListener
from ch.aplu.jcardgame.Hand import SortType
from enum import enum

# ------------------ class BitsCardListener --------------------------------
class BitsCardListener(CardListener):
   def atTarget(self, card, loc):
      card.getHand().draw()

# ------------------ class HandsCardListener -------------------------------
class HandsCardListener(CardListener):
   def __init__(self, i):
      self.i = i

   def leftDoubleClicked(self, card):
      card.setVerso(False)
      card.transfer(bids[self.i], True)
   
   def rightClicked(self, card):
      transferToStock(self.i)

# ------------------ global functions --------------------------------------
def dealingOut():
   global hands
   global talon
   cg.setStatusText("Dealing out...")
   hands = deck.dealingOut(nbPlayers, 0)  # All cards in talon
   talon = hands[nbPlayers]
   talon.setVerso(True)

   for i in range(nbPlayers):
      hands[i].setView(cg, StackLayout(handLocations[i], 90 * i))
      hands[i].draw()
      hands[i].setTouchEnabled(True)

   talon.setView(cg, StackLayout(handLocations[nbPlayers]))
   talon.draw()

   while not talon.isEmpty():
      for i in range(nbPlayers):
         top = talon.getLast()
         if i == 0:
            top.setVerso(False)
         else:
            top.setVerso(True)
         talon.transfer(top, hands[i], True)

def initBids():
   for i in range(nbPlayers):
      bids[i] = Hand(deck)
      bids[i].setView(cg, StackLayout(bidLocations[i]))
      bids[i].addCardListener(BitsCardListener())
      hands[i].setTargetArea(TargetArea(bidLocations[i]))
      hands[i].setTouchEnabled(True)
      hands[i].addCardListener(HandsCardListener(i))

def initStocks():
   for i in range(nbPlayers):
      stocks[i] = Hand(deck)
      if i == 0 or i == 2:
        rotationAngle = 0
      else:
        rotationAngle = 90
      stocks[i].setView(cg, StackLayout(stockLocations[i], rotationAngle))

def transferToStock(player):
   for i in range(nbPlayers):
      bids[i].setTargetArea(TargetArea(stockLocations[player]))
      card = bids[i].getLast()
      if card == None:
         continue
      card.setVerso(True)
      bids[i].transferNonBlocking(card, stocks[player], True)

# ------------------------ main --------------------------------------------
Suit = enum("SPADES", "HEARTS", "DIAMONDS", "CLUBS")
Rank = enum("ACE", "KING", "QUEEN", "JACK", "TEN")
deck = Deck(Suit.values(), Rank.values(), "cover");

nbPlayers = 4
handLocations = [Location(300, 525), Location(75, 300), Location(300, 75), Location(525, 300), Location(300, 300)]
bidLocations = [Location(300, 350), Location(250, 300), Location(300, 250), Location(350, 300)]
stockLocations = [Location(400, 500), Location(100, 400), Location(200, 100), Location(500, 200)]

hands = None
bids = [0] * nbPlayers
stocks = [0] * nbPlayers
talon = None
  
cg = CardGame(600, 600, 30)
dealingOut()
cg.setStatusText("Dealing out...done. Double click to play.")
initBids()
initStocks()
